core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -221,6 +221,211 @@ private[spark] class BlockManager(
+      blockSize: Long,
+      blockId: BlockId,
+      level: StorageLevel,
+      classTag: ClassTag[T],
+      tellMaster: Boolean,
+      keepReadLock: Boolean) {
+
+    /**
+     *  Reads the block content into the memory. If the update of the block store is based on a
+     *  temporary file this could lead to loading the whole file into a ChunkedByteBuffer.
+     */
+    protected def readToByteBuffer: ChunkedByteBuffer
+
+    protected def getInputStream(): InputStream
+
+    protected def saveToDiskStore(): Unit
+
+    private def saveDeserializedValuesToMemoryStore(inputStream: InputStream) = {
+      val values =
+        serializerManager.dataDeserializeStream(blockId, inputStream)(classTag)
+      memoryStore.putIteratorAsValues(blockId, values, classTag) match {
+        case Right(_) => true
+        case Left(iter) =>
+          // If putting deserialized values in memory failed, we will put the bytes directly
+          // to disk, so we don't need this iterator and can close it to free resources
+          // earlier.
+          iter.close()
+          false
+      }
+    }
+
+    private def saveSerializedValuesToMemoryStore(bytes: ChunkedByteBuffer) = {
+      val memoryMode = level.memoryMode
+      memoryStore.putBytes(blockId, blockSize, memoryMode, () => {
+        if (memoryMode == MemoryMode.OFF_HEAP && bytes.chunks.exists(!_.isDirect)) {
+          bytes.copy(Platform.allocateDirectBuffer)
+        } else {
+          bytes
+        }
+      })
+    }
+
+    /**
+     * Put the given data according to the given level in one of the block stores, replicating
+     * the values if necessary.
+     *
+     * If the block already exists, this method will not overwrite it.
+     *
+     * If keepReadLock is true, this method will hold the read lock when it returns (even if the
+     * block already exists). If false, this method will hold no locks when it returns.
+     *
+     * @return true if the block was already present or if the put succeeded, false otherwise.
+     */
+     def save(): Boolean = {
+      doPut(blockId, level, classTag, tellMaster, keepReadLock) { info =>
+        val startTimeMs = System.currentTimeMillis
+
+        // lazy as for a storage level which only targets disk (without any replication)
+        // this val won't be used at all
+        lazy val byteBuffer = readToByteBuffer
+
+        // Since we're storing bytes, initiate the replication before storing them locally.
+        // This is faster as data is already serialized and ready to send.
+        val replicationFuture = if (level.replication > 1) {
+          Future {
+            // This is a blocking action and should run in futureExecutionContext which is a cached
+            // thread pool. The ByteBufferBlockData wrapper is not disposed of to avoid releasing
+            // buffers that are owned by the caller.
+            replicate(blockId, new ByteBufferBlockData(byteBuffer, false), level, classTag)
```

Comment:
This is another place where you can save memory. Instead of allocating a `ByteBufferBlockData`, it could vary depending on the underlying storage. `DiskBlockData` and `EncryptedBlockData` are much cheaper here, as far as memory usage goes.

Reply:
And this way we can get rid of the `lazy val` too as instantiating `DiskBlockData` and `EncryptedBlockData` does not lead to loading the file content.